### ISSUE TITLE
attached image in ipynb not shown in html

### ISSUE THREAD

**Issue by abchenson** [jupyter/nbconvert#699]
[image_test.ipynb.zip](https://github.com/jupyter/nbconvert/files/1435173/image_test.ipynb.zip)
I tried converting this simple notebook with
`python -m nbconvert docs/notebooks/image_test.ipynb`
running nbconvert 5.3.1 . The notebook contains one markdown cell with a single line
`![image.png](attachment:image.png)`
But the image does not show up in the resulting html, only a cell with text 'image.png' in it.
 (attachment:image.png).  The image works in the notebook so it seems correctly embedded.  Is this not supported?  It's an important feature for me so any help would be really useful.
Thanks in advance

**Comment by platise** [comment]
I would also need this feature, as atm it is not possible to make any other output but browser's print to PDF.

For PDF export I would suggest images are exported under some temp dir, like .ipynb_tmp to be able to generate a PDF.

**Comment by platise** [comment]
How to increase intention for this feature?

Image attachments is such a cool feature with easy drag and drop images, makes it really possible to work in hardware development test and measurement projects, but without a proper HTML or PDF creation its use is limited.

**Comment by sglyon** [comment]
I needed this, so I wrote a pre-processor to extract attached images and put them in the `outputs`, which is used for images in the output part of a code cell.

It has been working well for me.

https://gist.github.com/sglyon/5687b8455a0107afc6f4c60b5f313670

**Comment by platise** [comment]
Thank you @sglyon for response.

Sounds interesting, and if I understand it: it extracts all images in dict{} which you then save into files, and use the files when generating the pdf with pdflatex?

If you have a small (one figure) example of usage it would be really helpful to understand the whole procedure and how you run it.

**Comment by sglyon** [comment]
No problem.

My usage was as follows

```python
from traitlets.config import Config
import nbformat
import nbconvert


def to_pdf(path)
    c = Config()
    my_preprocesors = [ExtractAttachmentsPreprocessor]
    c.PDFExporter.preprocessors = my_preprocesors
    c.LatexExporter.preprocessors = my_preprocesors

    c.Exporter.filters = {"markdown_filters": markdown_filters}
    exporter = nbconvert.PDFExporter(config=c, extra_loaders=[dl])
    exporter.template_file = "my_report.tplx"
    writer = nbconvert.writers.FilesWriter()

    resources = dict(
        output_files_dir="book_files",
        metadata=dict(name="Valorum training workshop"))

    big = nbformat.read(

    (body, resources) = exporter.from_notebook_node(big, resources=resources)
    writer.write(body, resources, "output.pdf")

```

**Comment by fuglede** [comment]
For what it's worth, fixing up the HTML post-conversion manually is pretty straightforward; assuming that all attachment filenames are unique, you could do something like

```python

with open('my_notebook.ipynb') as nb_file:
    nb_contents = nb_file.read()

# Convert using the ordinary exporter
notebook = nbformat.reads(nb_contents, as_version=4)
exporter = nbconvert.HTMLExporter()
body, res = exporter.from_notebook_node(notebook)

# Create a dict mapping all image attachments to their base64 representations
images = {}
for cell in notebook['cells']:
    if 'attachments' in cell:
        attachments = cell['attachments']
        for filename, attachment in attachments.items():
            for mime, base64 in attachment.items():
                images[f'attachment:{filename}'] = f'data:{mime};base64,{base64}'

# Fix up the HTML and write it to disk
for src, base64 in images.items():
    body = body.replace(f'src="{src}"', f'src="{base64}"')
with open('my_notebook.html', 'w') as output_file:
    output_file.write(body)
```

Of course, this does nothing to actually fix the bug. Without being able to claim much of an overview of how rendering occurs, it seems like you would want to somehow take this into account in the filter [`nbconvert.html.HTMLExporter.markdown2html`](https://github.com/jupyter/nbconvert/blob/99085946a98b3789dc4ce37e37a870b4d023c8f1/nbconvert/exporters/html.py#L69) (which does see the `NotebookNode` itself, but if you don't mind adding this as state, you could do so in `from_notebook_node`).

**Comment by platise** [comment]
Thank you @fuglede it is a great work-around export lib for the HTML output!

**Comment by platise** [comment]
Wrapper it together so other can imm download and use:

    Usage: ipy2html.py filename.ipynb [--slides]

[ipy2html.py.gz](https://github.com/jupyter/nbconvert/files/1809786/ipy2html.py.gz)

**Comment by imcomking** [comment]
Hello guys. Thanks for your workaround snippets.

I slightly change @platise's code which can handle multiple attachement.

``` python3
import nbformat
import nbconvert
import sys

if len(sys.argv) < 2:
    print("Usage:", sys.argv[0], 'filename.ipynb', '[--slides]')
    exit(-1)

with open(sys.argv[1]) as nb_file:
    nb_contents = nb_file.read()

# Convert using the ordinary exporter
notebook = nbformat.reads(nb_contents, as_version=4)
if len(sys.argv) == 3 and sys.argv[2] == '--slides':
    outname = sys.argv[1].split('.ipynb')[0] + '.slides.html'
    print("Converting to slides:", outname)    
    exporter = nbconvert.SlidesExporter()    
else:
    outname = sys.argv[1].split('.ipynb')[0] + '.html'
    print("Converting to HTML:", outname)
    exporter = nbconvert.HTMLExporter()
    
body, res = exporter.from_notebook_node(notebook)

# Create a list saving all image attachments to their base64 representations
images = []
for cell in notebook['cells']:
    if 'attachments' in cell:
        attachments = cell['attachments']
        for filename, attachment in attachments.items():
            for mime, base64 in attachment.items():
                images.append( [f'attachment:{filename}', f'data:{mime};base64,{base64}'] )

# Fix up the HTML and write it to disk
for itmes in images:
    src = itmes[0]
    base64 = itmes[1]
    body = body.replace(f'src="{src}"', f'src="{base64}"', 1)
    
with open(outname, 'w') as output_file:
    output_file.write(body)
```

The usage is same.

**Comment by imcomking** [comment]
Hello guys. Thanks for your workaround code snippets.

I slightly changed @platise's code which can handle multiple non-unique attachments.

``` python3
import nbformat
import nbconvert
import sys

if len(sys.argv) < 2:
    print("Usage:", sys.argv[0], 'filename.ipynb', '[--slides]')
    exit(-1)

with open(sys.argv[1]) as nb_file:
    nb_contents = nb_file.read()

# Convert using the ordinary exporter
notebook = nbformat.reads(nb_contents, as_version=4)
if len(sys.argv) == 3 and sys.argv[2] == '--slides':
    outname = sys.argv[1].split('.ipynb')[0] + '.slides.html'
    print("Converting to slides:", outname)    
    exporter = nbconvert.SlidesExporter()    
else:
    outname = sys.argv[1].split('.ipynb')[0] + '.html'
    print("Converting to HTML:", outname)
    exporter = nbconvert.HTMLExporter()
    
body, res = exporter.from_notebook_node(notebook)

# Create a list saving all image attachments to their base64 representations
images = []
for cell in notebook['cells']:
    if 'attachments' in cell:
        attachments = cell['attachments']
        for filename, attachment in attachments.items():
            for mime, base64 in attachment.items():
                images.append( [f'attachment:{filename}', f'data:{mime};base64,{base64}'] )

# Fix up the HTML and write it to disk
for itmes in images:
    src = itmes[0]
    base64 = itmes[1]
    body = body.replace(f'src="{src}"', f'src="{base64}"', 1)
    
with open(outname, 'w') as output_file:
    output_file.write(body)

```

The usage is same.

[ipy2html.py ](https://gist.github.com/imcomking/625d41730fb94f2d176a9c903175e880/raw/224aea9bba58be3d20bff15a773c53b46fd167bc/ipy2html.py)

**Comment by tonycpsu** [comment]
For what it's worth, I've had success with the [Export HTML with Embedded Images](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/export_embedded/readme.html) extension.  Not sure if there's a motivating use case that's not handled by it, but it works for me.

**Comment by platise** [comment]
Thanks @imcomking for upgrade, works well! and also handles exports where Embdded HTML doesn't work, with ref to @tonycpsu

**Comment by carlthome** [comment]
Status on this?

**Comment by boffi** [comment]
The proposed workarounds are fine but afaict `nbviewer` has no use for them and images embedded via MarkDown are rendered, e.g.,  with the string  `![My beautiful washing machine](drawing.svg)` displayed directly in the rendered notebook.

**Comment by Nikolai-Hlubek** [comment]
Thank you @imcomking and @platise for publishing your workarounds. They work fine. 

Ultimately it would be nice, if nbconvert could implement this functionality as well.

**Comment by MSeal** [comment]
We're always open to a nice PR :) I probably don't have time to move the work-around in and test it thoroughly for a while. But I'm trying to may attention to new PR contributions and not let them linger indefinitely.

**Comment by mgeier** [comment]
I tried the example at the top of this page with `master`, and it worked as expected.

So I guess this is resolved?
I'm not sure if it's in a release yet, though.

**Comment by maartenbreddels** [comment]
For reference, this is partly solved in #780 indeed, but i'm working on a fix and possible improvement in #980

**Comment by antoinecarme** [comment]
Nice to have.

**Comment by t-makaro** [comment]
This issue appears resolved by #980 (#1057 notwithstanding). However, this issue does appear to still occur for the PDFExporter, but I will open a separate issue for that.
